feat(anthropic): add programmatic tool calling

aayush-kapoor opened the pull request:
## Background

Anthropic released programmatic tool calling. It allows calling a client-side function tool from an ongoing provider-executed code execution tool.

The main assumption this breaks is that all tools are finished at the end of a turn. Special provider-executed tools may span turns.

See https://platform.claude.com/docs/en/agents-and-tools/tool-use/programmatic-tool-calling

## Summary

- Updated spec to account to add anthropic specific providerOptions for tools
- generate-text and stream-text core functions were modified to allow passing container IDs from previous steps (container ID is generated in the model response and needs to be passed back for this functionality)
- flag added specifically for the code_execution tool so that deferred tool results can be handled properly

## Manual Verification

- [x] unit tests
- [x] static examples
- [x] UI examples

## Checklist

- [x] Tests have been added / updated (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] A _patch_ changeset for relevant packages has been added (for bug fixes / features - run `pnpm changeset` in the project root)

## Related Issues

Fixes #10594

aayush-kapoor committed:
first commit

aayush-kapoor committed:
cs

aayush-kapoor committed:
Merge branch 'main' into aayush/anthropic-progr-tool-calling

aayush-kapoor committed:
updated unit tests

aayush-kapoor committed:
added docs

aayush-kapoor committed:
failing UI examples on multiturn

aayush-kapoor committed:
Merge branch 'main' into aayush/anthropic-progr-tool-calling

aayush-kapoor committed:
error handling

aayush-kapoor committed:
Merge branch 'main' into aayush/anthropic-progr-tool-calling

aayush-kapoor committed:
caller field passed for tool_use block

aayush-kapoor committed:
spec additions + UI working

aayush-kapoor committed:
rm

aayush-kapoor committed:
bot fix

aayush-kapoor committed:
Merge branch 'main' into aayush/anthropic-progr-tool-calling

lgrammel commented on examples/next-openai/components/tool/anthropic-code-execution-view.tsx:
```diff
@@ -159,7 +197,31 @@ function InputView({
+  }
+
+  // Handle programmatic tool calling format: { code: string }
+  if ('code' in input && !('type' in input)) {
```
this is strange. can we instead inject a fake type 'programmatic-tool-call' in our mapping?

lgrammel commented on packages/ai/src/generate-text/generate-text.ts:
```diff
@@ -499,13 +507,27 @@ A function that attempts to repair a tool call that failed to parse.
+                        },
+                      }),
+                      { ...providerOptions },
+                    )
```
looks strange

lgrammel commented on packages/provider-utils/src/types/tool.ts:
```diff
@@ -237,6 +237,19 @@ The ID of the tool. Must follow the format `<provider-name>.<unique-tool-name>`.
         args: Record<string, unknown>;
+
+        /**
+         * Whether this provider-executed tool supports deferred results.
```
defaults to false

lgrammel commented on packages/provider-utils/src/provider-tool-factory.ts:
```diff
@@ -75,10 +75,20 @@ export function createProviderToolFactoryWithOutputSchema<
   inputSchema: FlexibleSchema<INPUT>;
   outputSchema: FlexibleSchema<OUTPUT>;
+  /**
+   * Whether this provider-executed tool supports deferred results.
```
default to false

lgrammel commented on packages/ai/src/generate-text/generate-text.ts:
```diff
@@ -1009,10 +1065,43 @@ function asContent<TOOLS extends ToolSet>({
+
+            if (!supportsDeferredResults) {
+              throw new Error(`Tool call ${part.toolCallId} not found.`);
+            }
```
should we have this check? unsure

lgrammel commented on packages/ai/src/generate-text/stream-text.ts:
```diff
@@ -1231,6 +1236,20 @@ class DefaultStreamTextResult<TOOLS extends ToolSet, OUTPUT extends Output>
+                }),
+                { ...providerOptions },
+              )
+            : providerOptions;
```
maybe extract and test

lgrammel replied:
#11301

lgrammel committed:
Merge branch 'main' into aayush/anthropic-progr-tool-calling

aayush-kapoor committed:
Merge branch 'main' into aayush/anthropic-progr-tool-calling

lgrammel committed:
programmatic tool calling improvements (#11301)

## Summary

Improvements for #11262

## Tasks

- [x] introduce `forwardAnthropicContainerIdFromLastStep`
- [x] use `mergeObjects` for merging provider options
- [ ] add multi turn example

---------

Co-authored-by: Aayush Kapoor <83492835+aayush-kapoor@users.noreply.github.com>

aayush-kapoor committed:
Merge branch 'main' into aayush/anthropic-progr-tool-calling

aayush-kapoor committed:
Merge branch 'main' into aayush/anthropic-progr-tool-calling

aayush-kapoor committed:
streaming fix

aayush-kapoor committed:
ui example changed

aayush-kapoor committed:
js doc

aayush-kapoor committed:
proxy tool type added

aayush-kapoor committed:
test update

aayush-kapoor committed:
fixtures

aayush-kapoor committed:
fixture update

aayush-kapoor committed:
unit tests added

aayush-kapoor committed:
Merge branch 'main' into aayush/anthropic-progr-tool-calling

lgrammel committed:
rm unused
